lib/ProductOpener/Cache.pm

```diff
@@ -46,7 +48,37 @@ $memd = Cache::Memcached::Fast->new(
 	{
 		'servers' => $memd_servers,
 		'utf8' => 1,
+		compress_threshold => 10000,
 	}
 );
 
+my $json = JSON->new->utf8->allow_nonref->canonical;
+
+=head1 FUNCTIONS
+
+=head2  generate_cache_key($context_ref)
+
+Generate a key to use for caching, that depends on the content of the $context_ref object.
+
+=head3 Arguments
+
+=head4 $object_ref Reference to all the context / parameters etc. that have an influence on what we want to cache
+
+=head3 Return values
+
+MD5 of the key.
+
+=cut
+
+sub generate_cache_key ($context_ref) {
+
+	# We generate a sorted JSON so that we always have the same key for the context object
+	# even if it contains hashes (Storable::freeze may not have the same order of keys)
+	my $key = $server_domain . "/" . $json->encode($context_ref);
+	my $md5_key = md5_hex($key);
+	$log->debug("generate_cache_key", {context_ref => $context_ref, key => $key, md5_key => $md5_key})
+		if $log->is_debug();
+	return $md5_key;
```

Comment:
I find it a bit dangerous. Collision may happens.
They are not maybe a big issue if we at least prefix the key by - a domain + a function.

So I would associate each domain with a single character, same for each function, and then encode the rest of $context_ref with md5.

I'm not sure for the domain prefix for there might be too much case to handler, don't we have a variable saying the "product-family" ?  

```perl
SERVER_DOMAINS_PREFIX = {"openfoodfacts.org" => 'a', …}

FN_PREFIX = {"suggestion_cache" => 'a', "mongo_query" => 'b'}

sub generace_cache_key($fn, $context_ref) {
   $key = SERVER_DOMAINS_PREFIX->{$server_domain} . FN_PREFIX->{$fn};
   ...
   $key .= $md5_key
}
```

Reply:
I see below that you did it indeed, by adding a prefix… but in this case, let's add the prefix as a parameter to this function, it will be more clear.

And ok then for the domain in the hash (the most important thing is to have a "result type" prefix.

Reply:
ok, I made the key server domain + result type + hash of context, directly in the generate_cache_key function